### ISSUE TITLE
OpenTESArena not runing Linux

### ISSUE THREAD

**Issue by scapegoat01** [afritz1/OpenTESArena#141]
Hello, so today, I discovered that I could play Elder Scrolls Arena on Linux, so I did it. it all worked until the final command,  
> ./run.sh

 After I ran this command, my terminal outputed the following:

> ./TESArena: error while loading shared libraries: libsndio.so.6.1: cannot open shared object file: No such file or directory

No mater what I do, it does not work. The current os I am on is Parrot GNU/Linux 4.5 x86_64, host is a Latitude E6440 00, and the kernel is  4.19.0-parrot1-13t-amd64, command ran on a Mate-Terminal.

**Comment by afritz1** [comment]
Hi. I'm not sure what libsndio is. Presumably it's a dependency of SDL2 on Linux. No one else has reported this before so I don't know what the problem is, sorry. Maybe you need to download that shared library and install it somewhere.

**Comment by scapegoat01** [comment]
@afritz1 Honnestly, i dont know how that works

**Comment by afritz1** [comment]
I'm on Windows so I can't experiment with these things myself right now. Try googling a solution or wait for another Linux user to help. You can try asking on the project's Discord channel. The link is in the Readme.

**Comment by scapegoat01** [comment]
i think i found it

**Comment by afritz1** [comment]
What's the status of this? Did you figure out the issue?

**Comment by abelsromero** [comment]
If it serves, I reproduced the issue using the 0.8 release package.
Deleting the lib directory and using my installed libs fixed it.

**Comment by afritz1** [comment]
Hmm, so the distributed lib folder is making the Linux binaries not work out of the box? Are we using the wrong version of libraries, or are we not distributing enough libraries, or something else? @Ragora

**Comment by Translator5** [comment]
> "Sndio is a small audio and MIDI framework" @ http://www.sndio.org

I could not find the requested version "0.6.1" but installing "sndio v. 20180120-1" provides your "/usr/lib/libsndio.so.7.0" what you can copy to "OpenTESarena/libs" and rename to "libsndio.so.6.1".
This workaround is dirty and not recommended for long-term use, but it works to get the game to launch. 

But maybe this version-difference may cause the issues in my log; [gist](https://gist.github.com/Translator5/6bc38155cef19d13287e3712445a8f47#file-gistfile1-txt).

To the issues in the log, overall sound does work properly, but drops and doors work well.

**Comment by scapegoat01** [comment]
Sorry I haven't responded much, it's currently the exam period, so i have to study a lot, plus I've been kind of inactive on things like github, and other.

**Comment by scapegoat01** [comment]
in the end, i think ill just install it on windows

**Comment by scapegoat01** [comment]
i have installed it on windows

**Comment by afritz1** [comment]
Update on this? I'm still not sure if the distributed libraries in Linux releases are working for everyone.

**Comment by Babouchot** [comment]
Actually I think you are distributing an old version of libopenal. 
I just built and downloaded the last version you uploaded. 
The one I built is running right out of the box without any issue. 
The one I downloaded has the same issue as above. This is because the libopenal you distributed links against libsndio : 
```bash
$ ldd libopenal.so.1 
	linux-vdso.so.1 (0x00007ffe729af000)
	libsndio.so.6.1 => not found
	librt.so.1 => /usr/lib/librt.so.1 (0x00007f67e519f000)
	libpthread.so.0 => /usr/lib/libpthread.so.0 (0x00007f67e517e000)
	libdl.so.2 => /usr/lib/libdl.so.2 (0x00007f67e5179000)
	libm.so.6 => /usr/lib/libm.so.6 (0x00007f67e5033000)
	libc.so.6 => /usr/lib/libc.so.6 (0x00007f67e4e70000)
	/usr/lib64/ld-linux-x86-64.so.2 (0x00007f67e54a4000)
``` 

The one on my system does not : 
```bash
$ ldd /usr/lib/libopenal.so.1
	linux-vdso.so.1 (0x00007ffda33e6000)
	librt.so.1 => /usr/lib/librt.so.1 (0x00007f918c489000)
	libpthread.so.0 => /usr/lib/libpthread.so.0 (0x00007f918c468000)
	libdl.so.2 => /usr/lib/libdl.so.2 (0x00007f918c463000)
	libm.so.6 => /usr/lib/libm.so.6 (0x00007f918c31d000)
	libc.so.6 => /usr/lib/libc.so.6 (0x00007f918c15a000)
	/usr/lib64/ld-linux-x86-64.so.2 (0x00007f918c5ee000)
```

I suspect you probably use Ubuntu to build your linux version which usually has older versions of the same libraries than more "cutting edge distros" (I'm currently running on Arch linux).
I'm speculating here but some distros may have deprecated libsndio from their packages and this now build OpenAL without linking it to it. I'm saying this because this library is nowhere to be found on my system right now and that I have a package from the community named `libsndio-61-compat` that probably just install libsndio.6.1 (61), the one linked against by your libopenal.

But basically the script `run.sh` you did to run the game is overriding the libraries loading order by setting the LD_LIBRARY_PATH variable. Just running `./TESArena` will make the executable run without picking your bundled libraries but the system ones, and at least for me it makes the game run without issues. So this can be a short term fix for people having this issue (provided you have the other bundled libraries on your system : libSDL2-2.0.so.0 and libWildMidi.so.2)

I think the best fix you can do is to stop bundling the libopenal with the project, because I think that it should be installed on almost any distro that has a desktop installed anyway (this should be tested on several distros to ensure that). 
But to be sure you can add the installation of the OpenAL lib in your README install instruction. 
I think that it should be `sudo apt install libopenal` or something like that on an ubuntu system.

**Comment by Babouchot** [comment]
Otherwise, if you absolutely need support for libsndio within openal, you will need to also bundle the libsndio.so.6.1 along with its own dependencies, if any. Which doesn't seem to be the case since Lutris distribute it with its integrated Ubuntu-18.04-x86_64 runtime  and other programs like everspace or scummvm and it does not link against anything else that's not in the system: 
```bash
~/.local/share/lutris/runtime/Ubuntu-18.04-x86_64$ ldd libsndio.so.6.1 
	linux-vdso.so.1 (0x00007ffe7bffb000)
	librt.so.1 => /usr/lib/librt.so.1 (0x00007f868c32a000)
	libasound.so.2 => /usr/lib/libasound.so.2 (0x00007f868c248000)
	libbsd.so.0 => /usr/lib/libbsd.so.0 (0x00007f868c030000)
	libc.so.6 => /usr/lib/libc.so.6 (0x00007f868be6d000)
	libpthread.so.0 => /usr/lib/libpthread.so.0 (0x00007f868be4c000)
	libm.so.6 => /usr/lib/libm.so.6 (0x00007f868bd06000)
	libdl.so.2 => /usr/lib/libdl.so.2 (0x00007f868bcff000)
	/usr/lib64/ld-linux-x86-64.so.2 (0x00007f868c5a0000)
```
I copied it into my OpenTESArena libs folder and it now works fine with the run.sh launch script.

**Comment by afritz1** [comment]
The Linux builds have been getting built by @Ragora, and we thought that distributing those shared libraries like `libopenal.so.1` with the releases would allow for the greatest compatibility. If people have problems then they should build the engine themselves (paraphrased).

I've purged some libraries on my Ubuntu computer as part of figuring things out with @Ragora, so I don't really have a control group handy. But basically this is what I get from the Linux64 release of 0.10.0 (hand-copying from my Linux bash since I'm on Windows right now):
```bash
$ ldd ./TESArena
...
libopenal.so.1 => not found
libSDL2-2.0.so.0 => not found
libWildMidi.so.2 => not found
...
```
and ./TESArena doesn't work because it says it's missing libopenal.so.1. If I do ./run.sh then the game runs fine.
So with your proposed solution, I should run:
```
sudo apt-get install libopenal
```
to supposedly get it back to 'default' and then remove libopenal.so.1 and libsndio.so.6.1 from the libs folder? I'm not sure we can assume that everyone has the right libopenal on their computer. I don't know a whole lot about Linux libraries, sorry 😕

**Comment by Babouchot** [comment]
Ok, I thought that OpenAL could be bundled with the system, which apparently is not. 

So for your executable to run right out of the box, since you already distribute the libSDL2 and WildMidi dependencies, I think the simplest solution is to also provide the missing libsndio.so.6.1 in the libs folder. As I said several applications distribute it with their own executable, and given the upper output of ldd libsndio.so.6.1, I think we can safely assume that libsndio does not have further dependencies.
The easiest way to test that is with a fresh install preferably of another distro that the one you built from. I will test with a fresh VM of Fedora.

Linux libraries are not that much different from windows libraries, the only tricky thing is that not all distros bundle the same libraries by default and that you have to chase most of them in your system paths since you can install a lot of development libraries from your package manager. But as you have to do on windows, if you want the executable to run correctly you just have to provide your dependencies, and the ones of your dependencies recursively until the only dependencies left are the one bundled by the target system, which is relatively more difficult to find out on linux which is a bit less unified than windows.

**Comment by Babouchot** [comment]
Ok I just tested with your 0.10.0 and it works fine on my arch system now without any tweaks required. But it does not on my Fedora VM because of another library libXss.so but I just tested live and did not even installed it so I think I will test again later when I'll have some time with a real install, but that may be related to Fedora defaulting to Wayland as a display server instead of X11.

I noticed there is the issue #162 requesting a flatpak build. That could be a solution, flatpak aims at solving that kind of issues by providing common runtimes across linux distributions and that could be a way to provide a "it just works" executable along with an executable with minimal dependencies shipped like the one you are doing right now, and let the user choose. 
But I think the version you have right now is not far from working on any linux machine, the libsndio issue is fixed for me and the missing libXss on Fedora is a dependency from SDL2, so I'm pretty sure this is not that big of a deal since SDL2 runs properly on pretty much anything that has a CPU ^^. I'll check that out.
